### PR TITLE
Implement ScBasset

### LINKED ISSUE #1729: Implement scbasset
This meta-issue is for tracking the implementation of scbasset

- [ ] #1730 
- [ ] Implement module and basic model
- [ ] Add more model features

### LINKED ISSUE #1730: Add function to add sequences from fasta to anndata for atac-seq
https://github.com/calico/scBasset/blob/489e323e0942c823f603cbd5b4664f8bc49e6d9a/scbasset/utils.py#L20

https://muon.readthedocs.io/en/latest/api/generated/muon.atac.tl.html

```python
def add_dna_sequence_to_adata(
    adata: AnnData, 
    fasta_file: Path, 
    seq_len: int, 
    chr_var_key: str = "chr",
    start_var_key: str = "start",
    end_var_key: str = "end",
    sequence_varm_key="sequence", 
    code_varm_key="code"
) -> None:
    # Process dataframe of regions by seq_len
    # Make a dataframe of codes (e.g., A = 0...)
```

This can go in `scvi.data`

### PR DESCRIPTION
**Pull request by adamgayoso**
- Add data preprocessing to add dna sequences to adata (needs a test)
- Implement model and module
- Fix custom optimizer in training plan
- Fix categorical var field

TODO:

- [x] Add test for preprocessing function
- [x] Create a tutorial that reproduces some of their results (https://github.com/scverse/scvi-tutorials/pull/86), merge it on the tutorials repo and update the submodule here
- [x] User guide (can be simple like [solo](https://docs.scvi-tools.org/en/stable/user_guide/models/solo.html))
- [x] Hook up batch correction
- [x] All other todos in the code

Original code can be found [here](https://github.com/calico/scBasset/blob/main/scbasset/utils.py#L177).

If anyone would like to help, please comment on this PR and feel free to make a PR into this branch

Fixes #1729, Fixes #1730

See https://github.com/scverse/scvi-tools/issues/1840 for more TODOs in the future

### CODE REVIEW

File: scvi/data/_preprocessing.py [hunk 1]
```diff
@@ -351,3 +354,101 @@ def _concat_anndata(multi_anndata, other):
...
+    adata: anndata.AnnData,
+    seq_len: int = 1334,
+    genome_name: str = "hg38",
+    genome_dir: Optional[Path] = None,
+    genome_provider: str = "UCSC",
+    chr_var_key: str = "chr",
+    start_var_key: str = "start",
+    end_var_key: str = "end",
+    sequence_varm_key="dna_sequence",
+    code_varm_key="dna_code",
```
Comment (martinkim0) [inline scvi/data/_preprocessing.py:385]:
```suggestion
    sequence_varm_key: str = "dna_sequence",
    code_varm_key: str = "dna_code",
```

File: scvi/external/scbasset/_module.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+            sliced_seq = _seq[..., :-shift:]
+            return torch.concat([pad, sliced_seq], dim=-1)
+
+        def _shift_left(_seq):
+            # shift is negative
+            sliced_seq = _seq[..., -shift:]
+            return torch.concat([sliced_seq, pad], dim=-1)
+
+        sseq = _shift_right(seq) if shift > 0 else _shift_left(seq)
+        sseq.reshape(input_shape)
```
Comment (martinkim0) [inline scvi/external/scbasset/_module.py:166]: Would `F.pad` not work here?

File: scvi/external/scbasset/_module.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+    def __init__(self, shift_max=0, pad="uniform", **kwargs):
+        super().__init__()
+        self.shift_max = shift_max
+        self.register_buffer(
+            "augment_shifts", torch.arange(-self.shift_max, self.shift_max + 1)
+        )
+        self.pad = pad
+
+    def forward(self, seq_1hot: torch.tensor):
```
Comment (martinkim0) [inline scvi/external/scbasset/_module.py:124]:
```suggestion
    def forward(self, seq_1hot: torch.Tensor):
```

File: scvi/external/scbasset/_module.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+            if shift != 0:
+                sseq_1hot = self.shift_sequence(seq_1hot, shift)
+            else:
+                sseq_1hot = seq_1hot
+            return sseq_1hot
+        else:
+            return seq_1hot
+
+    @staticmethod
+    def shift_sequence(seq: torch.tensor, shift: torch.tensor, pad_value: float = 0.25):
```
Comment (martinkim0) [inline scvi/external/scbasset/_module.py:137]:
```suggestion
    def shift_sequence(seq: torch.Tensor, shift: torch.Tensor, pad_value: float = 0.25):
```

File: scvi/external/scbasset/_module.py (hunk 2 above)
Comment (adamgayoso) [reply scvi/external/scbasset/_module.py:166]: hmmm or https://pytorch.org/docs/stable/generated/torch.roll.html

File: scvi/external/scbasset/_module.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    ):
+        super().__init__()
+        self.conv = nn.Conv1d(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=kernel_size,
+            padding="same",
+            bias=False,
+        )
+        self.batch_norm = nn.BatchNorm1d(out_channels) if batch_norm else nn.Identity()
```
Comment (martinkim0) [inline scvi/external/scbasset/_module.py:43]: Looks like the default `eps` in keras is `0.001` while its `0.00001` for pytorch.

https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html

https://keras.io/api/layers/normalization_layers/batch_normalization/

Comment (adamgayoso) [reply scvi/external/scbasset/_module.py:43]: I'm going to leave all of this for another PR as I'm pretty sure all the linear layer initializations are different from the authors original code

File: scvi/external/scbasset/_module.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+        def _shift_right(_seq):
+            # shift is positive
+            sliced_seq = _seq[..., :-shift:]
+            return torch.concat([pad, sliced_seq], dim=-1)
+
+        def _shift_left(_seq):
+            # shift is negative
+            sliced_seq = _seq[..., -shift:]
+            return torch.concat([sliced_seq, pad], dim=-1)
```
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/external/scbasset/_module.py]: I went with roll and fill value

**Review by martinkim0 — APPROVED** [review]
